src/graph.cpp

```diff
@@ -531,6 +543,257 @@ static const char *s_monthNames[12] = { "Jan", "Feb", "Mar", "Apr", "May", "Jun"
+	switch (m_g.FontFace)
+	{
+	case 1: font = wxMetroTheme::Font(wxMT_LIGHT); break;
+	case 2: font = *wxSWISS_FONT; break;
+	case 3: font = wxFont(12, wxFONTFAMILY_ROMAN, wxFONTSTYLE_NORMAL, wxFONTWEIGHT_NORMAL); break;
+	case 4: font = wxFont(12, wxFONTFAMILY_MODERN, wxFONTSTYLE_NORMAL, wxFONTWEIGHT_NORMAL); break;
+	}
+
+	if (m_g.FontScale != 0.0)
+	{
+		int points = (int)(((double)font.GetPointSize()) * m_g.FontScale);
+		if (points < 4) points = 4;
+		if (points > 32) points = 32;
+		font.SetPointSize(points);
+	}
+
+	SetFont(font);
+
+	ShowGrid(m_g.CoarseGrid, m_g.FineGrid);
+	SetTitle(m_g.Title);
+	ShowLegend(m_g.ShowLegend);
+	SetLegendLocation((wxPLPlotCtrl::LegendPos)m_g.LegendPos);
+
+	// setup data
+	std::vector<VarValue*> yvars, xvars;
+	wxArrayString ynames, xnames;
+	int ndata = -1;
+
+	if (m_g.Type == Graph::CONTOUR)
+	{// TODO ensure m_g.X.count == 2 m_g.Y.count == 1
+		if (m_g.X.size() == 2 && m_g.Y.size() == 1) {
+			// detemine matrix size x[0] x x[1] = num sims
+			std::vector<double> xv, yv;
+			for (size_t i = 0; i < sims.size(); i++) {
+				double x = sims[i]->GetValue(m_g.X[0])->Value();
+				double y = sims[i]->GetValue(m_g.X[1])->Value();
+				auto itx = std::find(xv.begin(), xv.end(), x);
+				auto ity = std::find(yv.begin(), yv.end(), y);
+				if (itx == xv.end())
+					xv.push_back(x);
+				if (ity == yv.end())
+					yv.push_back(y);
+			}
+			// check for valid setup
+			if (xv.size() == 1 || yv.size() == 1) {
+				wxMessageBox("Invalid Contour plot setup, one independent variable has only one value");
+				return -1;
+			}
+			if (xv.size() * yv.size() != sims.size()) {
+				wxMessageBox("Invalid Contour plot setup, the product of the number of the two independent variable values is not equal to the number of parametric runs.");
+				return -1;
+			}
+
+			double zmin = 1e99, zmax = -1e99;
+			wxMatrix<double> XX, YY, ZZ;
+			size_t nx = xv.size(), ny = yv.size();
+			XX.Resize(ny , nx );
+			YY.Resize(ny,  nx);
+			ZZ.Resize(ny , nx );
+			for (size_t j = 0; j < ny; j++)
+			{
+				for (size_t i = 0; i < nx; i++)
+				{
+					XX.At(j , i ) = sims[i]->GetValue(m_g.X[0])->Value();
+					YY.At(j , i ) = sims[j*nx+i]->GetValue(m_g.X[1])->Value();
+					ZZ.At(j , i ) = sims[j * nx + i]->GetValue(m_g.Y[0])->Value();
+					if (ZZ.At(j , i ) < zmin) zmin = ZZ.At(j , i );
+					if (ZZ.At(j , i ) > zmax) zmax = ZZ.At(j , i );
+				}
+			}
+			wxPLContourPlot* plot = 0;
+			wxPLColourMap* jet = new wxPLParulaColourMap(zmin, zmax);
+			plot = new wxPLContourPlot(XX, YY, ZZ, true, wxEmptyString, 24, jet);
+			SetTitle(m_g.Title);
+			if (plot != 0)
+			{
+				AddPlot(plot, wxPLPlotCtrl::X_BOTTOM, wxPLPlotCtrl::Y_LEFT, wxPLPlotCtrl::PLOT_TOP, false);
+				SetSideWidget(jet);
+			}
+			//GetYAxis1()->SetReversed(true); // need setting
+			GetYAxis1()->SetLabel(m_g.YLabel);
+			GetXAxis1()->SetLabel(m_g.XLabel);
+			
+		}
+		ndata = sims.size();
+		return 0;
+
+	}
+	else
+	{ // TODO ensure m_g.X.count == 1 m_g.Y.count == 1
```

Comment:
Is this still relevant? It looks like vv->Length() is acceptable later.

Reply:
The one and one check is the check to plot output variable versus input variable for single outputs that are not monthly or timeseries based. An artifact of the GraphCtrl class in SAM that is a wrapper with limited functionality of PlotCtrl widget in wex so that autograph functions in the autograph.lk script can be used to show graphs on the results page. An opportunity to upgrade or expand some basic functionality in SAM.